### PR TITLE
Add basic networking functionality

### PR DESCRIPTION
**Pull request by srhickma**

### CODE REVIEW

**Review by WilsonCWu — COMMENTED** [review]
lgtm, attached @George-E and @AideTechBot to give a quick read over so they can approve

**Comment by WilsonCWu** [comment]
seems to run fine on my computer, got two players to run fine on my computer

**Comment by WilsonCWu** [comment]
@srhickma no rush to do it now, but i think you might need to get one of the mobile builds so the ui boxes render properly, since mobile has much higher pixel density and iphones have round corners on the edges :'(

**Comment by srhickma** [comment]
@WilsonCWu This is definitely not ready to merge, still lots of work that needs to be done. I just put up the pr to track it.

**Comment by WilsonCWu** [comment]
I think it might be worthwhile to merge now before the bulk of coding tmr, i feel like this would be really helpful for the game dev half even if its just the character that's synced

**Comment by srhickma** [comment]
> 
> 
> I think it might be worthwhile to merge now before the bulk of coding tmr, i feel like this would be really helpful for the game dev half even if its just the character that's synced

Makes sense, I will try to get the conflicts dealt with quick and merge, unless anyone else wants to review.

**Comment by srhickma** [comment]
@WilsonCWu Conflicts resolved, you should be able to merge whenever.

File: Assets/Scripts/RedRunner/Utilities/CameraController.cs [hunk 1]
```diff
@@ -78,8 +78,18 @@ void Update ()
 			}
 		}
 
+        public void Follow(Transform followee)
```
Comment (George-E) [inline Assets/Scripts/RedRunner/Utilities/CameraController.cs]: Nit: tabs/spaces mismatch

File: Assets/Scripts/RedRunner/Utilities/CameraController.cs [hunk 2]
```diff
@@ -78,8 +78,18 @@ void Update ()
...
+        {
+            m_Followee = followee;
+        }
+
 		public void Follow ()
 		{
+            if (m_Followee == null)
```
Comment (George-E) [inline Assets/Scripts/RedRunner/Utilities/CameraController.cs]: Nit: tabs/spaces mismatch

File: Assets/Scripts/RedRunner/Characters/RedCharacter.cs [hunk 3]
```diff
@@ -263,21 +270,39 @@ public override AudioSource Audio
 
 		void Awake ()
 		{
-			m_InitialPosition = transform.position;
-			m_InitialScale = transform.localScale;
-			m_GroundCheck.OnGrounded += GroundCheck_OnGrounded;
-			m_Skeleton.OnActiveChanged += Skeleton_OnActiveChanged;
+            m_InitialPosition = transform.position;
```
Comment (George-E) [inline Assets/Scripts/RedRunner/Characters/RedCharacter.cs]: Nit: tabs/spaces mismatch in this file

File: Assets/Scripts/RedRunner/Characters/RedCharacter.cs [hunk 4]
```diff
@@ -263,21 +270,39 @@ public override AudioSource Audio
...
+            base.OnStartLocalPlayer();
 
-		void Update ()
+            Local = this;
+            LocalPlayerSpawned();
+        }
+
+        void Update ()
 		{
-			if ( !GameManager.Singleton.gameStarted || !GameManager.Singleton.gameRunning )
```
Comment (George-E) [inline Assets/Scripts/RedRunner/Characters/RedCharacter.cs:280]: Why did you get rid of this check?

File: Assets/Scripts/RedRunner/GameManager.cs [hunk 5]
```diff
@@ -128,6 +127,13 @@ void Awake()
                 m_HighScore = 0f;
             }
 
+            RedCharacter.LocalPlayerSpawned += () =>
+            {
+                RedCharacter.Local.IsDead.AddEventAndFire(UpdateDeathEvent, this);
+                m_StartScoreX = RedCharacter.Local.transform.position.x;
+
+                StartGame();
```
Comment (George-E) [inline Assets/Scripts/RedRunner/GameManager.cs]: nice. makes sense

File: Assets/Scripts/RedRunner/TerrainGeneration/TerrainGenerator.cs [hunk 6]
```diff
@@ -156,7 +154,14 @@ public virtual void Generate ()
 				{
 					newX = 0f;
 				}
-				if ( block != null && ( m_LastBlock == null || newX < m_Character.transform.position.x + m_GenerateRange ) )
+
```
Comment (George-E) [inline Assets/Scripts/RedRunner/TerrainGeneration/TerrainGenerator.cs:157]: Nit: tabs/spaces mismatch

**Review by George-E — APPROVED** [review]
Lgtm. Left some small comments

File: Assets/Scripts/RedRunner/Utilities/CameraController.cs (hunk 1 above)
Comment (srhickma) [reply Assets/Scripts/RedRunner/Utilities/CameraController.cs]: Oh gross, does this project use tabs? I have mine set to 4 spaces.

File: Assets/Scripts/RedRunner/Characters/RedCharacter.cs (hunk 4 above)
Comment (srhickma) [reply Assets/Scripts/RedRunner/Characters/RedCharacter.cs:280]: The player no longer exists in the scene when the game starts. We only spawn the player once you join the network after hitting play, so both these checks will always be true.

**Pull request merged**